File: leetcode/string/1647. Minimum Deletions to Make Character Frequencies Unique/py/main.py
import collections

# Time O(N)
# Space O(N)
class Solution:
    def minDeletions(self, s):
        count = collections.Counter(s)

        freq_counter = collections.Counter()

        for k in count.keys():
          freq_counter[count[k]] = freq_counter[count[k]] + 1

        res = 0
        
        for key in list(freq_counter):
          if freq_counter[key] > 1:
            freq = freq_counter[key]

            for _ in range(freq - 1):
              rest_cnt = key

              while rest_cnt >= 0 and rest_cnt in freq_counter:
                res += 1
                rest_cnt -= 1
              
              if rest_cnt > 0:
                freq_counter[rest_cnt] = 1

        return res

File: leetcode/string/1647. Minimum Deletions to Make Character Frequencies Unique/js/main.js
// Time O(N*K)
// Space O(N)
const minDeletions = s => {
  let map1 = {};

  for (let x of s) {
    map1[x] = ~~map1[x] + 1;
  }

  let map2 = {};

  for (let k of Object.keys(map1)) {
    map2[map1[k]] = ~~map2[map1[k]] + 1;
  }

  let cnt = 0;

  for (let k of Object.keys(map2)) {
    if (map2[k] > 1) {
      let freq = map2[k];

      for (let i = 0; i < freq - 1; i++) {
        let x = k;

        while (x >= 0 && x in map2) {
          cnt++;
          x--;
        }

        if (x > 0) {
          map2[x] = 1;
        }
      }
    }
  }

  return cnt;
};
